### PR TITLE
Jetpack connect: Implement plan interval routes

### LINKED ISSUE #12497: JPC: Add routes for monthly plans
We’d like routes that we can use to automatically add a monthly plan to cart instead of the yearly plans:

- https://wordpress.com/jetpack/connect/personal/monthly
- https://wordpress.com/jetpack/connect/premium/monthly
- https://wordpress.com/jetpack/connect/pro/monthly

These routes should look and work just like the ones _without_ the `monthly` suffix but instead of adding the _yearly_ plan to the user's cart, they should add the respective _monthly_ plan.

cc @bubel

### PR DESCRIPTION
**Pull request by sirreal**
Supersedes #14408
Contains #14416 
Fixes #12497

To test:

The goal of this PR is to unify the handlers so that plan type and interval logic can happen in one place.

I'm happy to discuss whether this is a move in the right direction, I've had a few false starts trying to simplify and clarify this part of the flow.

Combine several similar route handlers into one handler.
Explicitly match valid intervals `monthly` or `yearly`.

**To test:**
* It's a good idea to set the following for relevant debug information:
   ```js
   localStorage.setItem('debug', [
     'calypso:jetpack-connect:actions',
     'calypso:jetpack-connect:controller',
   ] )
   ```
* Test the following urls. You can append `?url=$site` to test a url directly, be sure to this functionality:
  * http://calypso.localhost:3000/jetpack/connect/install
  * http://calypso.localhost:3000/jetpack/connect/personal
  * http://calypso.localhost:3000/jetpack/connect/personal/yearly
  * http://calypso.localhost:3000/jetpack/connect/personal/monthly
  * http://calypso.localhost:3000/jetpack/connect/premium
  * http://calypso.localhost:3000/jetpack/connect/premium/yearly
  * http://calypso.localhost:3000/jetpack/connect/premium/monthly
  * http://calypso.localhost:3000/jetpack/connect/pro
  * http://calypso.localhost:3000/jetpack/connect/pro/yearly
  * http://calypso.localhost:3000/jetpack/connect/pro/monthly
* There should be no warnings/errors in the console
* When using a `(pro|premium|personal)/(monthly|yearly)` entry point, after connection you should be redirected to checkout with the expected plan (no plans page).
* Ensure tests pass:
   ```
   npm run test-client client/state/jetpack-connect/
   ```

The `/pro`, `/premium`, and `/personal` routes should complete the flow up at a checkout screen for the selected plan, _not_ the plans screen.

Preparation for #14408

### CODE REVIEW

**Comment by sirreal** [comment]
Ping @richardmuscat, implementation for `/monthly` and `/yearly` routes.

File: client/jetpack-connect/controller.js [hunk 1]
```diff
@@ -29,6 +29,14 @@ import { setDocumentHeadTitle as setTitle } from 'state/document-head/actions';
 import { getSelectedSiteId } from 'state/ui/selectors';
 import { isJetpackSite } from 'state/sites/selectors';
 import PlansLanding from './plans-landing';
+import {
+	PLAN_JETPACK_PREMIUM,
```
Comment (sirreal) [inline client/jetpack-connect/controller.js]: 😱  😭  What have I come to?! b9cc40ea054057556560bd1a50798cb1f0980753

File: client/jetpack-connect/controller.js [hunk 2]
```diff
@@ -42,6 +50,34 @@ const analyticsPageTitleByType = {
 	pro: 'Jetpack Install Pro',
 };
 
+/**
+ * Calculate plan based on type and interval.
```
Comment (sirreal) [inline client/jetpack-connect/controller.js]: 👍  b9cc40ea054057556560bd1a50798cb1f0980753

File: client/state/jetpack-connect/test/selectors.js [hunk 3]
```diff
@@ -7,6 +7,9 @@ import { expect } from 'chai';
  * Internal dependencies
  */
 import {
+	PLAN_JETPACK_PREMIUM,
```
Comment (sirreal) [inline client/state/jetpack-connect/test/selectors.js]: 👍  b9cc40ea054057556560bd1a50798cb1f0980753

File: client/jetpack-connect/plans.jsx [hunk 4]
```diff
@@ -5,30 +5,37 @@ import { connect } from 'react-redux';
...
+import { isFreePlan } from 'lib/plans';
 import { getPlansBySite } from 'state/sites/plans/selectors';
 import { recordTracksEvent } from 'state/analytics/actions';
 import { getCurrentUser } from 'state/current-user/selectors';
 import { addItem } from 'lib/upgrades/actions';
 import { selectPlanInAdvance, goBackToWpAdmin, completeFlow } from 'state/jetpack-connect/actions';
 import QueryPlans from 'components/data/query-plans';
 import QuerySitePlans from 'components/data/query-site-plans';
-import { isRequestingPlans, getPlanBySlug } from 'state/plans/selectors';
+import { isRequestingPlans, getPlanBySlugSelector } from 'state/plans/selectors';
```
Comment (sirreal) [inline client/jetpack-connect/plans.jsx]: 😅  ☕️  ☕️  Thanks for catching that! e714611bd0a82ca078cd457aa45b27dda33d3d5a

File: client/jetpack-connect/plans.jsx [hunk 5]
```diff
@@ -41,6 +48,15 @@ import { isSiteAutomatedTransfer } from 'state/selectors';
 const CALYPSO_REDIRECTION_PAGE = '/posts/';
 const CALYPSO_PLANS_PAGE = '/plans/my-plan/';
 const JETPACK_ADMIN_PATH = '/wp-admin/admin.php?page=jetpack';
+const JETPACK_PLANS = [
```
Comment (sirreal) [inline client/jetpack-connect/plans.jsx]: 👍  This was on my radar, thanks for the prodding. See #14755

File: client/jetpack-connect/controller.js [hunk 6]
```diff
@@ -42,6 +50,34 @@ const analyticsPageTitleByType = {
...
+ *
+ * @param   {String?} type     Flow type: personal | premium | pro
+ * @param   {String?} interval Interval: yearly | monthly
+ * @returns {String?}          Jetpack plan slug
+ */
+const calculatePlan = ( type, interval ) => {
+	if ( type === 'personal' ) {
```
Comment (sirreal) [inline client/jetpack-connect/controller.js]: Moved, refactored, and added tests in 3d4c7d5333608f72beb7b12909cec3e373bbe30b

**Comment by sirreal** [comment]
@tyxla Awesome review, thanks. I've addressed some things, and others will be adressed in other PRs, starting with #14755. Putting this on hold for now.

File: client/state/jetpack-connect/actions.js [hunk 7]
```diff
@@ -76,17 +76,19 @@ export default {
 		};
 	},
 
-	checkUrl( url, isUrlOnSites, flowType ) {
+	checkUrl( url, isUrlOnSites, selectedPlan ) {
 		return ( dispatch ) => {
 			if ( _fetching[ url ] ) {
 				return;
 			}
+			const action = {
```
Comment (sirreal) [inline client/state/jetpack-connect/actions.js]: Just to be clear, you're talking about creating a new function (`createCheckUrlAction`) in the same module that is this:

```js
const createCheckUrlAction = ( url, selectedPlan ) => ( {
	type: JETPACK_CONNECT_CHECK_URL,
	selectedPlan,
	url,
} );
```

Is that correct?

Comment (tyxla) [inline client/state/jetpack-connect/actions.js]: Yes, that's exactly what I meant. With one small difference: you'll need to export this function in order to add tests for it.

**Comment by sirreal** [comment]
Reported: 544-gh-jpop-issues

**Comment by apeatling** [comment]
Looks like this is on hold, please reopen when it's active again with new code or updates.

**Pull request closed without merging**